moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -30,25 +29,55 @@
  * @author Jesse Wilson
  */
 abstract class ClassFactory<T> {
+
+  /**
+   * This can't be a {@link LinkedHashTreeMap} since {@link Class}
+   * does not implement {@link Comparable}.
+   */
+  private static Map<Class<?>, ClassFactory<?>> unsafeFactories = new HashMap<>();
+
   abstract T newInstance() throws
       InvocationTargetException, IllegalAccessException, InstantiationException;
 
   public static <T> ClassFactory<T> get(final Class<?> rawType) {
+    if (unsafeFactories.containsKey(rawType)) {
```

Comment:
Just call get and check against null so you don't do hash and lookup steps twice.